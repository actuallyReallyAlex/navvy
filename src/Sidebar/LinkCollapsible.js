import React, { Fragment, useState } from 'react'
import PropTypes from 'prop-types'
import { Box, Button, Text } from 'grommet'
import { FormNext, FormDown } from 'grommet-icons'
import LinkIndicator from './LinkIndicator'
import LinkIcon from './LinkIcon'
import LinkCollapsibleChild from './LinkCollapsibleChild'

const LinkCollapsible = ({
  children,
  handler,
  icon,
  selected,
  setSelected,
  size,
  title
}) => {
  const [expanded, setExpanded] = useState(false)

  const handleClick = () => {
    setSelected(title)
    setExpanded(!expanded)
    handler && handler()
  }

  return (
    <Fragment>
      <Button hoverIndicator onClick={handleClick} plain>
        <Box
          background={expanded ? 'light-5' : undefined}
          direction="row"
          height="45px"
          margin={
            expanded
              ? { top: 'xsmall', bottom: 'none', left: 'xsmall' }
              : { vertical: 'xsmall', left: 'xsmall' }
          }
          pad={{ vertical: 'small', left: 'small' }}
          style={{ position: 'relative' }}
          width="95%"
        >
          {selected === title && <LinkIndicator />}
          <Box
            align="center"
            direction="row"
            gap="small"
            justify={size === 'expanded' ? 'start' : 'center'}
            width={size === 'expanded' ? '100%' : '75%'}
          >
            <LinkIcon icon={icon} />
            {size === 'expanded' && <Text>{title}</Text>}
          </Box>
          <Box align="center" justify="center" pad={{ horizontal: 'small' }}>
            {expanded ? <FormDown /> : <FormNext />}
          </Box>
        </Box>
      </Button>
      {expanded &&
        children.map(({ handler, icon, title }, i) => (
          <LinkCollapsibleChild
            handler={handler}
            key={i}
            icon={icon}
            size={size}
            title={title}
          />
        ))}
    </Fragment>
  )
}

LinkCollapsible.propTypes = {
  children: PropTypes.array.isRequired,
  handler: PropTypes.func,
  icon: PropTypes.string,
  selected: PropTypes.string.isRequired,
  setSelected: PropTypes.func.isRequired,
  size: PropTypes.oneOf(['expanded', 'collapsed']).isRequired,
  title: PropTypes.string.isRequired
}

export default LinkCollapsible
