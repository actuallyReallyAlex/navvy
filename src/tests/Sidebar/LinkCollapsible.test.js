import React from 'react'
import LinkCollapsible from '../../Sidebar/LinkCollapsible'
import { shallow } from 'enzyme'

const defaultProps = {
  children: [{ icon: 'Cube', title: 'Child 1' }],
  handler: jest.fn(),
  icon: 'Cube',
  selected: 'Selected',
  setSelected: jest.fn(),
  size: 'expanded',
  title: 'Title'
}

describe('<LinkCollapsible />', () => {
  afterEach(() => {
    defaultProps.handler.mockClear()
    defaultProps.setSelected.mockClear()
  })

  test('Should render the <LinkCollapsible /> component.', () => {
    const component = shallow(<LinkCollapsible {...defaultProps} />)
    expect(component).toMatchSnapshot()
  })

  test('Should call the link handler if the link is clicked.', () => {
    const component = shallow(<LinkCollapsible {...defaultProps} />)
    expect(defaultProps.handler).toHaveBeenCalledTimes(0)
    component.find('WithTheme(Button)').simulate('click')
    expect(component).toMatchSnapshot()
    expect(defaultProps.handler).toHaveBeenCalledTimes(1)
  })
})
