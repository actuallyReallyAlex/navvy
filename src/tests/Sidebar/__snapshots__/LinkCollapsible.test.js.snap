// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkCollapsible /> Should call the link handler if the link is clicked. 1`] = `ShallowWrapper {}`;

exports[`<LinkCollapsible /> Should render the <LinkCollapsible /> component as a non-selected Link with sidebar collapsed. 1`] = `ShallowWrapper {}`;

exports[`<LinkCollapsible /> Should render the <LinkCollapsible /> component as a non-selected Link with sidebar expanded. 1`] = `ShallowWrapper {}`;

exports[`<LinkCollapsible /> Should render the <LinkCollapsible /> component as a selected Link with sidebar collapsed. 1`] = `ShallowWrapper {}`;

exports[`<LinkCollapsible /> Should render the <LinkCollapsible /> component as a selected Link with sidebar expanded. 1`] = `ShallowWrapper {}`;
